http/client/request.go

```diff
@@ -38,10 +52,82 @@ func NewRequest(L *lua.LState) int {
 
 	req := &luaRequest{Request: httpReq}
 	req.Request.Header.Set(`User-Agent`, DefaultUserAgent)
-	ud := L.NewUserData()
-	ud.Value = req
-	L.SetMetatable(ud, L.GetTypeMetatable("http_request_ud"))
-	L.Push(ud)
+	L.Push(lvRequest(L, req))
+	return 1
+}
+
+// http.filerequest(url, files, params) returns user-data, error
+func NewFileRequest(L *lua.LState) int {
+	url := L.CheckString(1)
+	files := L.CheckTable(2)
+
+	body := new(bytes.Buffer)
+	writer := multipart.NewWriter(body)
+	var writeFile = func(info *lua.LTable, w *multipart.Writer) (err error) {
+		fieldname := info.RawGetString("fieldname")
+		path := info.RawGetString("path")
+		if fieldname == lua.LNil || path == lua.LNil {
+			return errors.New("fieldname or path is nil")
+		}
+		filename := info.RawGetString("filename")
+		if filename == lua.LNil {
+			filename = lua.LString(filepath.Base(path.String()))
+		}
+
+		part, err := writer.CreateFormFile(fieldname.String(), filename.String())
+		if err != nil {
+			return
+		}
+		file, err := os.Open(path.String())
+		if err != nil {
+			return
+		}
+		defer file.Close()
+		_, err = io.Copy(part, file)
+		return
+	}
+
+	var err error
+	if files.Len() == 0 {
+		err = writeFile(files, writer)
+	} else {
+		for key, value := files.Next(lua.LNil); key != lua.LNil; key, value = files.Next(key) {
+			err = writeFile(value.(*lua.LTable), writer)
+			if err != nil {
+				break
+			}
+		}
+	}
+	if err != nil {
+		L.Push(lua.LNil)
+		L.Push(lua.LString(err.Error()))
+		return 2
+	}
+
+	if L.GetTop() > 2 {
+		L.CheckTable(3).ForEach(func(k, v lua.LValue) {
+			writer.WriteField(k.String(), v.String())
```

Comment:
Hmm… does this need Next and error handling too? can `WriteField` really fail?

Reply:
Nudge - if you answer or correct this, I think I can approve and merge.

Reply:
Yes, I shouldn't ignore any errors.